### PR TITLE
Maintenance: use SBufs for ErrorDetailEntry data fields

### PR DESCRIPTION
**Pull request by kinkie**
Also fixed (re)configuration crashes when an error-details.txt entry is
missing a required "detail" or "descr" field.

### CODE REVIEW

File: src/HttpHeaderTools.cc
Comment (rousskov) [inline src/HttpHeaderTools.cc:242]: Most likely, the long-term "primary" function should have a different API. For example, I would expect it return an SBuf (and an exception thrown on errors, as usual). The closer you can get to that correct long-term interface in this PR, when you are adding the very first callers for that function, the better! Otherwise, we would be creating new technical debt (i.e. wrong HTTP quoted string parsing API) while trying to get rid of the old technical debt (i.e. String).

File: src/ssl/ErrorDetailManager.cc
Comment (rousskov) [inline src/ssl/ErrorDetailManager.cc:66]: Returning a pointer to a temporary local variable. Please address all such bugs.

**Review by rousskov — CHANGES_REQUESTED** [review]
Thank you for starting this conversion. It is more difficult than it may seem. I left a few comments in hope to help you make progress, but please do not overlook PR build failures because, I suspect, they may require non-trivial adjustments as well.

File: src/HttpHeaderTools.cc
Comment (kinkie) [reply src/HttpHeaderTools.cc:246]: Doing what you suggest would balloon the scope of the PR.

I would prefer an approach where
1. data members and simple interfaces get ported. There are 58 files including SquidString right now
2. once they are ported, get rid of the bridge methods like this one; one PR per method, building unit tests first to ensure non regressions

Comment (rousskov) [reply src/HttpHeaderTools.cc:246]: > Doing what you suggest would balloon the scope of the PR.

The above assertion is false because suggested option 1 keeps the scope of this PR exactly the same.

> I would prefer an approach where [we first introduce bridge functions and then get rid of them].

I am not against bridge functions per se, and suggested option 1 arguably continues to use a bridge function. A bridge function is the best way forward in many cases, but each suggested conversion should be evaluated on its own. Attempts to confine all future conversions to one conversion technique (e.g., bridge functions) seem counter productive to me because Squid code is too varied for one technique to be the right choice in all conversion cases.

As you know, I support minimal conversion PRs and even prefer those whenever they are feasible. Needless to say, there are many ways to convert and even many ways to create bridge functions. Naturally, there are certain principles or invariants we should observe _regardless_ of the PR size or the overall conversion approach. This change request is about violations of some of those principles: 

* avoid performance regressions (option 2 avoids such a regression but requires more work)
* if you have to create a performance regression, let the potential caller know (option 1 does that)

File: src/security/ErrorDetail.cc
Comment (rousskov) [inline src/security/ErrorDetail.cc:679]: [Same](https://github.com/squid-cache/squid/pull/1451/files#r1289104529) problem in this newly added change: Unlike the old condition, the new condition is always true.

File: src/HttpHeaderTools.cc
Comment (kinkie) [reply src/HttpHeaderTools.cc]: > > Doing what you suggest would balloon the scope of the PR.
> 
> The above assertion is false because suggested option 1 keeps the scope of this PR exactly the same.

You are right; I had improperly focused only on part of your answer.

> > I would prefer an approach where [we first introduce bridge functions and then get rid of them].
> 
> I am not against bridge functions per se, and suggested option 1 arguably continues to use a bridge function. A bridge function is the best way forward in many cases, but each suggested conversion should be evaluated on its own. Attempts to confine all future conversions to one conversion technique (e.g., bridge functions) seem counter productive to me because Squid code is too varied for one technique to be the right choice in all conversion cases.
> 
> As you know, I support minimal conversion PRs and even prefer those whenever they are feasible. Needless to say, there are many ways to convert and even many ways to create bridge functions. Naturally, there are certain principles or invariants we should observe _regardless_ of the PR size or the overall conversion approach. This change request is about violations of some of those principles:
> 
> * avoid performance regressions (option 2 avoids such a regression but requires more work)
> * if you have to create a performance regression, let the potential caller know (option 1 does that)

I'll for for option 1 for now. I would like to push through getting rid of SquidStrig

File: src/HttpHeaderTools.cc
Comment (kinkie) [reply src/HttpHeaderTools.cc]: Done. I had put it next to the current one but I have no strong preference

File: src/security/ErrorDetail.cc
Comment (kinkie) [reply src/security/ErrorDetail.cc:665]: I think we can get away with an isEmpty() test here. An uninitialised detailEntry will have a default-constructed SBuf which has zero length

File: src/HttpHeaderTools.cc
Comment (kinkie) [reply src/HttpHeaderTools.cc]: Done in the SBuf-based variant

File: src/HttpHeaderTools.cc
Comment (kinkie) [reply src/HttpHeaderTools.cc]: Some of these are borne out of wanting to avoid issues such as [this other comment](https://github.com/squid-cache/squid/pull/1451#discussion_r1289147284), coupled with the opportunity that these methods have few callers. I'll see where it leads and that it doesn't blow the scope up to much

Comment (rousskov) [reply src/HttpHeaderTools.cc]: > I'll see where it leads and that it doesn't blow the scope up to much

Sounds like a good plan to me. When assessing where these changes have led, please keep in mind the essence of my previous [comment](https://github.com/squid-cache/squid/pull/1451#discussion_r1291613871): The added long-term API changes will need to be adjusted before they can land -- there is currently a significant amount of expected additional work in this PR. triggered by those long-term API changes.

File: src/ssl/ErrorDetailManager.cc
Comment (kinkie) [reply src/ssl/ErrorDetailManager.cc]: I think using a static variable should be enough for that

File: src/ssl/ErrorDetailManager.cc
Comment (kinkie) [inline src/ssl/ErrorDetailManager.cc:63]: @rousskov , I must be missing something really important. Such as.. where is theList initialised?

File: src/ssl/ErrorDetailManager.cc
Comment (rousskov) [reply src/ssl/ErrorDetailManager.cc]: Using a static variable in such a context is unsafe because it silently screws up two callers in close proximity with each other. When possible, we should avoid such time bombs.

**Comment by kinkie** [comment]
commit 61fdffb makes it clear that an ErrorDetailEntry is either all set or not set,
and adds a way to test it, making intent clear in callers

File: src/security/ErrorDetail.cc
Comment (kinkie) [reply src/security/ErrorDetail.cc]: Addressed in 61fdffb

File: src/ssl/ErrorDetailManager.cc
Comment (kinkie) [reply src/ssl/ErrorDetailManager.cc]: What is the best way to do it? SBuf::copy() to a static buffer is very kludgy (e.g. it doesn't guarantee zero-termination)

File: src/security/ErrorDetail.cc
Comment (kinkie) [reply src/security/ErrorDetail.cc]: Addressed in commit 61fdffb

**Comment by kinkie** [comment]
@rousskov , can you check the status I'm at? Thanks!

**Comment by rousskov** [comment]
> @rousskov , can you check the status I'm at?

Sure, I flipped this PR to my side. Please give me a few days.

File: src/security/ErrorDetail.cc
Comment (rousskov) [inline src/security/ErrorDetail.cc:740]: I bet we are looping because the original code (before 83b053a) was looping and changing from a loop to a map lookup is not trivial. The original code had to loop because it used a raw/POD array.

File: src/security/ErrorDetail.cc
Comment (rousskov) [inline src/security/ErrorDetail.cc:727]: There are a lot of problems in this official code, and these problems are in the way of doing the Right Thing in this PR. I am too busy to fix this quickly myself. If you can help, I can suggest the next step: This method should take an std::ostream reference (instead of a pointer to a c-string). The error extractors in the map below should print their information into the supplied stream (instead of returning a string).

**Review by rousskov — DISMISSED** [review]

File: src/ssl/ErrorDetailManager.cc
Comment (rousskov) [reply src/ssl/ErrorDetailManager.cc]: > I hope to find the answer to that question when I resume working on this PR.

I do not have a complete/polished answer, but its beginnings are sketched at https://github.com/squid-cache/squid/pull/1451#discussion_r1297800747.

File: src/security/ErrorDetail.cc
Comment (rousskov) [reply src/security/ErrorDetail.cc]: > This method should take an std::ostream reference (instead of a pointer to a c-string). The error extractors in the map below should print their information into the supplied stream (instead of returning a string).

Done in 326b13c, but I have not reviewed, tested, and polished the result yet.

Comment (rousskov) [reply src/security/ErrorDetail.cc]: > Done in [326b13c](https://github.com/squid-cache/squid/commit/326b13c98494cca1d95241d052f6db7bf425f495), but I have not reviewed, tested, and polished the result yet.

I pushed more fixes (as of commit 51ba982), but I am not done yet.

File: src/ssl/ErrorDetailManager.cc
Comment (rousskov) [inline src/ssl/ErrorDetailManager.cc:35]: This check is missing in the official code. Without this check httpHeaderParseQuotedString() crashes in the unlikely event an error-details.txt entry is missing a required "detail" or "descr" field. I have disclosed this fix in the updated PR description.

httpHeaderParseQuotedString() also has parsing bugs. I do not think we should touch that old function in this PR. I also do not think we should adjust our SlowlyParseQuotedString() wrapper to compensate for some of those httpHeaderParseQuotedString() problems. The wrapper should remain a wrapper.

File: src/ssl/ErrorDetailManager.cc
Comment (rousskov) [inline src/ssl/ErrorDetailManager.cc:71]: I have removed getErrorDescr() and getErrorDetail() (and the corresponding ErrorDetailsManager methods) because they had to be modified to work with SBuf _and_ these "find record and extract the requested field" methods were only making things more complex. An updated findRecord() interface addresses old caller needs without requiring a wrapper function for every field and "forwarding" those wrappers through the ErrorDetailsManager class to the ultimate consumer.

The diff does not show some of these modifications well, unfortunately.

**Review by rousskov — COMMENTED** [review]
@kinkie, I am done with adjusting this PR. I hope the current version is OK, but there is a lot of noise, and I could have easily missed something. The ball is on your side.

File: src/security/ErrorDetail.h
Comment (kinkie) [reply src/security/ErrorDetail.h:87]: Removing the TODO. Hopefully over time we will rename things,
there is no point in singling anythign specific out

File: src/ssl/ErrorDetailManager.cc
Comment (kinkie) [reply src/ssl/ErrorDetailManager.cc:35]: okay, reverting, adding to my personal TODO list to look into httpHeaderParseQuotedString

File: src/ssl/ErrorDetailManager.cc
Comment (kinkie) [reply src/ssl/ErrorDetailManager.cc:71]: ok, thanks!

File: src/security/ErrorDetail.h
Comment (rousskov) [reply src/security/ErrorDetail.h:87]: > there is no point in singling anythign specific out

The point of this specific TODO is documented in this change request. Do you think we cannot agree on the new names? That is the precondition for that TODO removal. It is not clear (to me) whether that precondition has been met.

File: src/ssl/ErrorDetailManager.cc
Comment (kinkie) [reply src/ssl/ErrorDetailManager.cc:35]: Okay; I notice that Http::SlowlyParseQuotedString has some checking and uses exceptions for the fail path, so it seems compatible with the intent here. At this point, this wrapper is useless at all, I can just call Http::SlowlyParseQuotedString, as this function only has two callsites.

File: src/ssl/ErrorDetailManager.cc
Comment (rousskov) [inline src/ssl/ErrorDetailManager.cc:34]: Calling hasNamed() without checking its result does not make much sense. Before we decide how to write this code, could you please explain what was the motivation behind commit 088a048c92fd49daa1606b686edfd613a39c0d09? Commit message does not answer that question AFAICT.

I am aware of the following two statements, but they do not help me answer that "why commit https://github.com/squid-cache/squid/commit/088a048c92fd49daa1606b686edfd613a39c0d09" question:

>> Alex at https://github.com/squid-cache/squid/pull/1451#discussion_r1337658245: This check is missing in the official code. Without this check httpHeaderParseQuotedString() crashes in the unlikely event an error-details.txt entry is missing a required "detail" or "descr" field. I have disclosed this fix in the updated PR description.

>> httpHeaderParseQuotedString() also has parsing bugs. I do not think we should touch that old function in this PR. I also do not think we should adjust our SlowlyParseQuotedString() wrapper to compensate for some of those httpHeaderParseQuotedString() problems. The wrapper should remain a wrapper.

> Francesco at https://github.com/squid-cache/squid/pull/1451#discussion_r1347716925: Okay; I notice that Http::SlowlyParseQuotedString has some checking and uses exceptions for the fail path, so it seems compatible with the intent here. At this point, this wrapper is useless at all, I can just call Http::SlowlyParseQuotedString, as this function only has two callsites.

File: src/security/ErrorDetail.cc
Comment (rousskov) [inline src/security/ErrorDetail.cc:752]: This renaming is rather misleading because this method does not convert ErrorCode (yes, there is such a thing in this context) to its description. This code expands `%code`. I recommend leaving this method name unchanged in this PR.

If (and only if) you insist on renaming, I can suggest:

```suggestion
Security::ErrorDetail::expandPercentCode(const char * const code, std::ostream &os) const
```

or

```suggestion
Security::ErrorDetail::substitutePercentCode(const char * const code, std::ostream &os) const
```

or something like that...

**Review by rousskov — CHANGES_REQUESTED** [review]

File: src/security/ErrorDetail.cc
Comment (kinkie) [reply src/security/ErrorDetail.cc:752]: I like expandPercentCode; "convert" is very  nondescriptive

File: src/ssl/ErrorDetailManager.cc
Comment (kinkie) [reply src/ssl/ErrorDetailManager.cc:34]: hasNamed() is misnamed; the third argument is an OUT parameter, which returns into the provided string, maybe this is the source of the confusion. Can we do without checking the return value? Yes, because if the header is not found, an empty string will be returned, and Http::SlowlyParseQuotedString will call httpHeaderParseQuotedString which will fail because the first character is not a ```"``` .

I had interpreted your request in this way; but yes, if you ask me, one more free check would help

Comment (kinkie) [reply src/ssl/ErrorDetailManager.cc]: I misinterpreted your comment as a suggestion to remove the check. Accepted your suggestion reinstating it, I'm good with the code as is. I don't think there's any more pending requests for improvements, are there?

Comment (rousskov) [reply src/ssl/ErrorDetailManager.cc]: > I don't think there's any more pending requests for improvements, are there?

I do not see any. Please feel free to clear for merge when you are comfortable with PR code.

**Comment by rousskov** [comment]
> @rousskov , ready at any time; it's only missing an explicit approval

This PR was missing an `M-cleared-for-merge` label. It does not need an explicit approval (because your implicit approval is enough to merge this PR on a "slow burner" path).

FYI: It takes Anubis a few minutes to scan all PRs after a manual PR change, so you will see a stale "PR approval" check right after manual PR changes.

As of now, if all PR tests succeed and no new merging problems (or last minute objections) surface, then this PR should be automatically staged and merged.

**Pull request closed without merging**
